### ISSUE TITLE
apoc.custom.list displays procedure name but execution errors out with `There is no procedure with the name`

### ISSUE THREAD

**Issue by ckazz** [neo4j-contrib/neo4j-apoc-procedures#1568]
## Expected Behavior
After deploying a new APOC custom procedure I expect it to be found and executed.

## Actual Behavior
For full stract trace please see below:
```
2020-06-16 13:32:24.502+0000 ERROR [o.n.b.v.r.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.General.UnknownError]: There is no procedure with the name `custom.copy_course` registered for this database instance. Please ensure you've spelled the procedure name correctly and that the procedure is properly deployed., reference aeb0d3fa-4e8a-4491-8556-a8573aa8477d. There is no procedure with the name `custom.copy_course` registered for this database instance. Please ensure you've spelled the procedure name correctly and that the procedure is properly deployed.
org.neo4j.graphdb.QueryExecutionException: There is no procedure with the name `custom.copy_course` registered for this database instance. Please ensure you've spelled the procedure name correctly and that the procedure is properly deployed.
at apoc.custom.CypherProcedures$CustomStatementRegistry$2.apply(CypherProcedures.java:231)
at org.neo4j.kernel.impl.proc.ProcedureRegistry.callProcedure(ProcedureRegistry.java:219)
at org.neo4j.kernel.impl.proc.Procedures.callProcedure(Procedures.java:290)
at org.neo4j.kernel.impl.newapi.AllStoreHolder.callProcedure(AllStoreHolder.java:1098)
at org.neo4j.kernel.impl.newapi.AllStoreHolder.procedureCallWrite(AllStoreHolder.java:884)
at org.neo4j.cypher.internal.runtime.interpreted.CallSupport$$anonfun$4.apply(CallSupport.scala:73)
at org.neo4j.cypher.internal.runtime.interpreted.CallSupport$$anonfun$4.apply(CallSupport.scala:73)
at org.neo4j.cypher.internal.runtime.interpreted.CallSupport$.callProcedure(CallSupport.scala:167)
at org.neo4j.cypher.internal.runtime.interpreted.CallSupport$.callReadWriteProcedure(CallSupport.scala:74)
at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext.callReadWriteProcedure(TransactionBoundQueryContext.scala:977)
at org.neo4j.cypher.internal.compatibility.v3_5.ExceptionTranslatingQueryContext$$anonfun$callReadWriteProcedure$1.apply(ExceptionTranslatingQueryContext.scala:169)
at org.neo4j.cypher.internal.compatibility.v3_5.ExceptionTranslatingQueryContext$$anonfun$callReadWriteProcedure$1.apply(ExceptionTranslatingQueryContext.scala:169)
at org.neo4j.cypher.internal.compatibility.v3_5.ExceptionTranslationSupport$class.translateException(ExceptionTranslationSupport.scala:33)
at org.neo4j.cypher.internal.compatibility.v3_5.ExceptionTranslatingQueryContext.translateException(ExceptionTranslatingQueryContext.scala:42)
at org.neo4j.cypher.internal.compatibility.v3_5.ExceptionTranslationSupport$class.translateIterator(ExceptionTranslationSupport.scala:47)
at org.neo4j.cypher.internal.compatibility.v3_5.ExceptionTranslatingQueryContext.translateIterator(ExceptionTranslatingQueryContext.scala:42)
at org.neo4j.cypher.internal.compatibility.v3_5.ExceptionTranslatingQueryContext.callReadWriteProcedure(ExceptionTranslatingQueryContext.scala:169)
at org.neo4j.cypher.internal.runtime.interpreted.DelegatingQueryContext.callReadWriteProcedure(DelegatingQueryContext.scala:253)
at org.neo4j.cypher.internal.runtime.EagerReadWriteCallMode.callProcedure(ProcedureCallMode.scala:66)
at org.neo4j.cypher.internal.runtime.interpreted.pipes.ProcedureCallPipe.org$neo4j$cypher$internal$runtime$interpreted$pipes$ProcedureCallPipe$$call(ProcedureCallPipe.scala:90)
at org.neo4j.cypher.internal.runtime.interpreted.pipes.ProcedureCallPipe$$anon$$$$e6882d5096c8b8441ac36d3c0cc459e$$$$ateResultsByAppending$1.apply(ProcedureCallPipe.scala:73)
at org.neo4j.cypher.internal.runtime.interpreted.pipes.ProcedureCallPipe$$anon$$$$e6882d5096c8b8441ac36d3c0cc459e$$$$ateResultsByAppending$1.apply(ProcedureCallPipe.scala:71)
at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:435)
at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:441)
at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
at org.neo4j.cypher.internal.runtime.RuntimeJavaValueConverter$feedQueryResultRecordIteratorToVisitable.accept(RuntimeJavaValueConverter.scala:63)
at org.neo4j.cypher.internal.compatibility.v3_5.runtime.PipeExecutionResult.accept(PipeExecutionResult.scala:74)
at org.neo4j.cypher.internal.compatibility.v3_5.runtime.executionplan.StandardInternalExecutionResult.materializeResult(StandardInternalExecutionResult.scala:77)
at org.neo4j.cypher.internal.compatibility.v3_5.runtime.executionplan.StandardInternalExecutionResult.initiate(StandardInternalExecutionResult.scala:59)
at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$initiate$1.apply$mcV$sp(ClosingExecutionResult.scala:65)
at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$initiate$1.apply(ClosingExecutionResult.scala:65)
at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$initiate$1.apply(ClosingExecutionResult.scala:65)
at org.neo4j.cypher.exceptionHandler$runSafely$.apply(exceptionHandler.scala:89)
at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult.org$neo4j$cypher$internal$compatibility$ClosingExecutionResult$$safely(ClosingExecutionResult.scala:167)
at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult.initiate(ClosingExecutionResult.scala:65)
at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$.wrapAndInitiate(ClosingExecutionResult.scala:214)
at org.neo4j.cypher.internal.compatibility.CypherCurrentCompiler$CypherExecutableQuery$$anonfun$execute$3.apply(CypherCurrentCompiler.scala:216)
at org.neo4j.cypher.internal.compatibility.CypherCurrentCompiler$CypherExecutableQuery$$anonfun$execute$3.apply(CypherCurrentCompiler.scala:190)
at org.neo4j.cypher.exceptionHandler$runSafely$.apply(exceptionHandler.scala:89)
at org.neo4j.cypher.internal.compatibility.CypherCurrentCompiler$CypherExecutableQuery.execute(CypherCurrentCompiler.scala:223)
at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:116)
at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:85)
at org.neo4j.bolt.v1.runtime.TransactionStateMachineV1SPI$BoltResultHandleV1.start(TransactionStateMachineV1SPI.java:189)
at org.neo4j.bolt.v1.runtime.TransactionStateMachine$State.startExecution(TransactionStateMachine.java:506)
at org.neo4j.bolt.v1.runtime.TransactionStateMachine$State$2.run(TransactionStateMachine.java:385)
at org.neo4j.bolt.v1.runtime.TransactionStateMachine.run(TransactionStateMachine.java:110)
at org.neo4j.bolt.v1.runtime.TransactionStateMachine.run(TransactionStateMachine.java:98)
at org.neo4j.bolt.v3.runtime.TransactionReadyState.processRunMessage(TransactionReadyState.java:82)
at org.neo4j.bolt.v3.runtime.TransactionReadyState.processUnsafe(TransactionReadyState.java:49)
at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48)
at org.neo4j.bolt.v1.runtime.BoltStateMachineV1.nextState(BoltStateMachineV1.java:144)
at org.neo4j.bolt.v1.runtime.BoltStateMachineV1.process(BoltStateMachineV1.java:92)
at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:89)
at org.neo4j.bolt.runtime.MetricsReportingBoltConnection.lambda$enqueue$0(MetricsReportingBoltConnection.java:68)
at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:191)
at org.neo4j.bolt.runtime.MetricsReportingBoltConnection.processNextBatch(MetricsReportingBoltConnection.java:86)
at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:139)
at org.neo4j.bolt.runtime.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:171)
at org.neo4j.bolt.runtime.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$2(ExecutorBoltScheduler.java:154)
at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
```

#### Procedure creation
```
CALL apoc.custom.asProcedure(
"copy_course",
"// Clone individual Course node and set the version
CALL apoc.log.info('Entry in copyCourse from courseId %d', [$courseId])
MATCH (course:Course)
WHERE ID(course) = $courseId
SET course.clonePairingId=course.externalId
WITH course as course
CALL apoc.refactor.cloneNodes([course], false, $skipProperties) YIELD output as clonedCourse
SET clonedCourse.version=$newCourseVersion
WITH course as course, clonedCourse as clonedCourse
// Create a COPIED relationship back to the original course
CREATE (clonedCourse)-[:COPIED {copyType:$copyType}]->(course)
WITH course as course, clonedCourse as clonedCourse
// Build Course Subgraph (excluding blacklisted nodes)
OPTIONAL MATCH path = (course)-[r1]-(level1:CourseContext)
WHERE TYPE(r1) <> 'COPIED'
AND NONE(label IN LABELS(level1) WHERE label IN $skipLabels)
WITH COLLECT(path) AS pathTotal, course as course, clonedCourse as clonedCourse
// Level 2 nodes
OPTIONAL MATCH path2 = (course)-[r1]-(level1:CourseContext)-[r2]-(level2:CourseContext)
WHERE TYPE(r1) <> 'COPIED' AND TYPE(r2) <> 'COPIED'
AND NONE(label IN LABELS(level1) WHERE label IN $skipLabels)
AND NONE(label IN LABELS(level2) WHERE label IN $skipLabels)
WITH pathTotal + COLLECT(path2) AS pathTotal2, course as course, clonedCourse as clonedCourse
// Level 3 nodes
OPTIONAL MATCH path3 = (course)-[r1]-(level1:CourseContext)-[r2]-(level2:CourseContext)-[r3]-(level3:CourseContext)
WHERE TYPE(r1) <> 'COPIED' AND TYPE(r2) <> 'COPIED' AND TYPE(r3) <> 'COPIED'
AND NONE(label IN LABELS(level1) WHERE label IN $skipLabels)
AND NONE(label IN LABELS(level2) WHERE label IN $skipLabels)
AND NONE(label IN LABELS(level3) WHERE label IN $skipLabels)
WITH pathTotal2 + COLLECT(path3) AS pathTotal3, course as course, clonedCourse as clonedCourse
// Clone the non-blacklisted Course Subgraph
CALL apoc.refactor.cloneSubgraphFromPaths(pathTotal3, {standinNodes:[[course, clonedCourse]], skipProperties:$skipProperties}) YIELD input, output, error
WITH DISTINCT clonedCourse as clonedCourse, course as course
// Create relationships from Course to CourseResource
MATCH (course)-[r:COURSE_AND_COURSE_RESOURCE]->(courseResource:CourseResource)
CREATE (clonedCourse)-[r2:COURSE_AND_COURSE_RESOURCE {createdBy:'COURSE_COPY', createdDate:timestamp()}]->(courseResource)
SET r2.order=r.order, r2.progenitorId=r.progenitorId
WITH DISTINCT course as course, clonedCourse as clonedCourse
CALL apoc.log.info('HERE! of copyCourse with source courseId %d.', [$courseId])
WITH DISTINCT course as course, clonedCourse as clonedCourse
// Create relationships from CourseResourceReferences to CourseResources
MATCH (course)-[]-(:CourseContext)-[]-(ref:CourseContext:CourseResourceReference)<-[r:COURSE_RESOURCE_AND_COURSE_RESOURCE_REFERENCE]-(courseResource:CourseResource)
WITH course as course, clonedCourse as clonedCourse, courseResource as courseResource, ref as ref, r as r
MATCH (clonedCourse)-[]-(:CourseContext)-[]-(ref2:CourseContext:CourseResourceReference) WHERE ref2.clonePairingId=ref.externalId
WITH DISTINCT course as course, clonedCourse as clonedCourse, courseResource as courseResource, ref as ref, ref2 as ref2, r as r
CREATE (courseResource)-[r2:COURSE_RESOURCE_AND_COURSE_RESOURCE_REFERENCE {createdBy:'COURSE_COPY', createdDate:timestamp()}]->(ref2)
SET r2.order=r.order, r2.progenitorId=r.progenitorId
WITH DISTINCT course as course, clonedCourse as clonedCourse
CALL apoc.log.info('HERE2! of copyCourse with source courseId %d.', [$courseId])
WITH DISTINCT course as course, clonedCourse as clonedCourse
// Create relationships from CourseResourceReferences to CourseResourceFormats
MATCH (course)-[]-(:CourseContext)-[]-(ref:CourseContext:CourseResourceReference)<-[r:COURSE_RESOURCE_FORMAT_AND_COURSE_RESOURCE_REFERENCE]-(crf:CourseResourceFormat)
WITH course as course, clonedCourse as clonedCourse, crf as crf, ref as ref, r as r
MATCH (clonedCourse)-[]-(:CourseContext)-[]-(ref2:CourseContext:CourseResourceReference) WHERE ref2.clonePairingId=ref.externalId
WITH DISTINCT course as course, clonedCourse as clonedCourse, crf as crf, ref as ref, ref2 as ref2, r as r
CREATE (crf)-[r2:COURSE_RESOURCE_FORMAT_AND_COURSE_RESOURCE_REFERENCE {createdBy:'COURSE_COPY', createdDate:timestamp()}]->(ref2)
SET r2.order=r.order, r2.progenitorId=r.progenitorId
WITH DISTINCT clonedCourse as clonedCourse
CALL apoc.log.info('Exit of copyCourse with source courseId %d.', [$courseId])
RETURN clonedCourse AS clonedCourse",
"write",
[['clonedCourse', 'NODE']],
[['courseId', 'LONG'], ['newCourseVersion', 'DOUBLE'], ['copyType', 'STRING'], ['skipProperties', 'LIST OF STRING'], ['skipLabels', 'LIST OF STRING']],
"Copies a given course by courseId including relationships and omitting properties as specified in skipProperties and labels as specified in skipLabels");
```

#### Procedure call
```
WITH
7631095 as courseId,
1.0 AS newCourseVersion,
'COURSE_COPY' AS copyType,
['guid', 'externalId', 'createdBy', 'createdDate', 'updatedBy', 'updatedDate'] AS skipProperties,
['Note', 'VersionCategory'] AS skipLabels
call custom.copy_course(courseId, newCourseVersion, copyType, skipProperties, skipLabels) YIELD clonedCourse
return clonedCourse;
```

#### Actual behavior
At first, execution if succeeds and then fails subsequent executions.

## Screenshots (where it's possibile)
Screenshot from neo4j browser 
![neo4j_browser](https://user-images.githubusercontent.com/46497511/84940654-017ff480-b095-11ea-8b4a-4c2623705e3f.png)

### Versions

  - OS: Windows 2010
  - Neo4j: 3.5.18 (single instance)
  - Neo4j-Apoc: 3.5.0.12

**Comment by conker84** [comment]
@ckazz  can you try with this jar?
[apoc-3.5.0.12-all.jar.zip](https://github.com/neo4j-contrib/neo4j-apoc-procedures/files/4920248/apoc-3.5.0.12-all.jar.zip)

**Comment by ckazz** [comment]
Thanks @conker84. Shared with the user facing the issue and will update you as soon as I hear back.

**Comment by ckazz** [comment]
@conker84 the jar did not resolve their issue. I'm asking whether they can update their findings directly here.

**Comment by conker84** [comment]
@jranua can you please share the md5 of the file?

**Comment by jranua** [comment]
@conker84 
MD5 hash of apoc-3.5.0.12-all.jar:
b6adfd51365bd9e26e78f1978fae0f22
CertUtil: -hashfile command completed successfully.

**Comment by conker84** [comment]
@jranua I don't know what you're are doing wrong but I have the same MD5 and it works fine.
can you try with a fresh instance?

**Comment by jranua** [comment]
Hmm, I'm thoroughly confused. I now have a stripped down Neo4j 3.5.18 with the apoc jar you provided. Originally I had a graphaware plugin also included. I am running on a Windows 10 VM. 

I re-deployed the procedure, made sure it was available (call apoc.custom.list) and then tried to run it. I purposely put a syntax error in the call and it correctly told me my data type was wrong (see screenshot). Once I correct the syntax error it fails with the same error.

Any ideas on things to try?
![custom_proc_not_found](https://user-images.githubusercontent.com/41159865/89229879-ec5d1780-d5a7-11ea-817c-d223ec98b116.png)

**Comment by jranua** [comment]
Do you have any updates? As I stated above, I tried the same process with a basic Neo4j 3.5.18 install with just the APOC 3.5.0.12 plugin.

**Comment by jranua** [comment]
Do all of the commit references imply you were able to recreate the issue? If so, do you have any idea when this would be fixed with a release of apoc 3.5?

**Comment by fbiville** [comment]
@jranua I merged the pull request https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/1778 yesterday.
You can build the `3.5` branch locally and test that the resulting APOC JAR fixes the issue.

**Comment by tygrysmalajski** [comment]
I'm having the same issue on Neo4j: 4.2.3 and Neo4j-Apoc: 4.2.0.1

**Comment by conker84** [comment]
closed via 1d6c032
